### PR TITLE
Product Gallery Thumbnails: Fix overflow issues and improve responsiveness

### LINKED ISSUE #11014: Product Gallery: Thumbnails layout and position need some attention
## Describe the bug

After https://github.com/woocommerce/woocommerce-blocks/pull/10867# we should re-review how the Thumbnails block displays in various scenarios.

For example:

269298815-b1a338df-3c49-4a04-b95a-5a7280bdf0b3

## To reproduce

1. With a block theme, open `Appearance > Editor`.
2. Open `Templates > Single Product`.
3. Select `Convert to Blocks` if you are using the Classic Templates.
4. Add the `Product Gallery` block.
5. Test the Thumbnails position in various setting configurations on the frontend.

### PR DESCRIPTION
**Pull request by danieldudzic**
- This PR attempts to improve the way Thumbnails display and re-size in the editor and the frontend, preventing them from overflowing the container.
- This PR also changes the minimum number of thumbnails from `2` to `3`.

## What

Fixes woocommerce/woocommerce-blocks#11014

## Why

Currently, the thumbnails have a hardcoded size and overflow the container.

## Important notes

1. The Thumbnails cropping needs to be fixed in a separate PR:
- Slack reference: p1697717112780469/1697706500.523259-slack-C02FL3X7KR6
- https://github.com/woocommerce/woocommerce-blocks/pull/11718
2. This PR **does not** fix the thumbnails styling in the dialog. That will need to be completely separate.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. With a block theme, open `Appearance > Editor`.
3. Open `Templates > Single Product`.
4. Select `Convert to Blocks` if you are using the Classic Templates.
5. Add the `Product Gallery` block.
6. Ensure the Thumbnails are enabled.
7. Test how thumbnails are displaying for various `Number of Thumbnails` values.
8. Test with thumbnails `Position` set to `off`, `left`, `bottom` and `top`.
9. Test both the editor and front end.

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

### Editor

| Before | After |
| ------ | ----- |
|before_editor|after_editor|

### Front end

| Before | After |
| ------ | ----- |
|before__frontend|after_frontend|

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Product Gallery Thumbnails: Fix overflow issues and improve responsiveness.

### CODE REVIEW

**Comment by roykho** [comment]
So this change breaks the styling when inside the dialog.



Also (not related to this PR). I feel the thumbnails should be a fixed size whether they are right/left/bottom position. In addition, I believe they're currently too big. I've had a look at other popular e-commerce sites, none of them have thumbnails this big.

**Comment by danieldudzic** [comment]
> So this change breaks the styling when inside the dialog.
> 
> 
> 
> Also (not related to this PR). I feel the thumbnails should be a fixed size whether they are right/left/bottom position. In addition, I believe they're currently too big. I've had a look at other popular e-commerce sites, none of them have thumbnails this big.

You are right, there's some wonkiness with a larger container width.

I guess this should be resolved by addressing:

> In this PR I was unable to completely remove the CSS based sizing of the Thumbnails container. This probably should be refactored in the future, as the user should be able to control the width of Product Gallery components using the Gutenberg width controls. I was unable to do so, without breaking the CSS animation of the Large Image.

> Also (not related to this PR). I feel the thumbnails should be a fixed size whether they are right/left/bottom position. In addition, I believe they're currently too big. I've had a look at other popular e-commerce sites, none of them have thumbnails this big.

If we hardcode the size there's no way for the user to adjust it.

I think it makes sense to adjust the thumbnail size to the container width. Patterns/users can use width controls to control the thumbnail size.

The bottom position thumbnail size adjusts to fit all thumbnails in one line in this PR, as we don't have a setting for the thumbnail size.

I have tried to find the best thumbnail solution based on the inspiration from this discussion: p1697706500523259-slack-C02FL3X7KR6 - but perhaps there's a different _preferred_ solution when min/max settings are not present.

Would you mind sharing your two cents on this, @jarekmorawski?

**Comment by jarekmorawski** [comment]
@danieldudzic and I had a chat about different ways to solve the sizing issue and we've decided to keep the scaling behavior while tweaking the layout when the thumbnails are on the right or left to always match the large image's size. We've also decided to limit the min. number of thumbnails to 4 or 3 and treat that as a reasonable min. size limit.

Here's a simple demo:

https://github.com/woocommerce/woocommerce-blocks/assets/79307566/cc1df4bf-69fa-46a0-a6f6-78880fcfdb32

We'll update the issue with any new information and pivot if the new approach doesn't work out.

**Comment by danieldudzic** [comment]
> @danieldudzic and I had a chat about different ways to solve the sizing issue and we've decided to keep the scaling behavior while tweaking the layout when the thumbnails are on the right or left to always match the large image's size. We've also decided to limit the min. number of thumbnails to 4 or 3 and treat that as a reasonable min. size limit.
> 
> Here's a simple demo:
> galleryscale.mov
> 
> We'll update the issue with any new information and pivot if the new approach doesn't work out.

I have now added changes to the thumbnails scaling when they are displayed vertically.

IGPU4QnKDf

File: tests/e2e/tests/product-gallery/inner-blocks/product-gallery-thumbnails/product-gallery-thumbnails.block_theme.side_effects.spec.ts [hunk 1]
```diff
@@ -60,22 +62,38 @@ test.describe( `${ blockData.name }`, () => {
...
-		const largeImageBlockBoundingRect = await largeImageBlock.boundingBox();
 
 		await expect( thumbnailsBlock ).toBeVisible();
-		// Check the default position: on the left of the large image
-		await expect( thumbnailsBlockBoundingRect?.y ).toBeGreaterThan(
-			largeImageBlockBoundingRect?.y as number
+
+		// We should refactor this.
+		// eslint-disable-next-line playwright/no-wait-for-timeout
+		await page.waitForTimeout( 500 );
```
Comment (gigitux) [inline tests/e2e/tests/product-gallery/inner-blocks/product-gallery-thumbnails/product-gallery-thumbnails.block_theme.side_effects.spec.ts:70]: If you will not fix it this PR, please create a follow-up issue!

**Review by gigitux — APPROVED** [review]
Amazing work! 👏 Everything LGTM. I want to highlight two things that I don't think are related to this PR:

There is a weird animation when the layout of the Product Gallery changes. I think that this is low priority:
https://github.com/woocommerce/woocommerce-blocks/assets/4463174/4f0ef79a-7ad6-468d-9d50-004f8eaf122d

Additionally, I observed that when a product contains only one image, we display the thumbnails:

image

During my previous involvement with this, there was a specific custom logic in place to address this situation.


Please, if you do not work on fixing the `waitForTimeout` in this PR, create a follow-up issue.

**Comment by danieldudzic** [comment]
> There is a weird animation when the layout of the Product Gallery changes. I think that this is low priority: https://github.com/woocommerce/woocommerce-blocks/assets/4463174/4f0ef79a-7ad6-468d-9d50-004f8eaf122d

This animation comes from Gutenberg. I suspect that it's a result of the `Stack` group block changing into a `Row` and the blocks moving up and down at the same time. Not sure if there's a way to suppress it. I'd also categorize it as low priority, but I have now opened an issue for it: https://github.com/woocommerce/woocommerce/issues/42061

> Additionally, I observed that when a product contains only one image, we display the thumbnails

I have now fixed that. Great catch!

File: tests/e2e/tests/product-gallery/inner-blocks/product-gallery-thumbnails/product-gallery-thumbnails.block_theme.side_effects.spec.ts (hunk 1 above)
Comment (danieldudzic) [reply tests/e2e/tests/product-gallery/inner-blocks/product-gallery-thumbnails/product-gallery-thumbnails.block_theme.side_effects.spec.ts:70]: Created! https://github.com/woocommerce/woocommerce-blocks/issues/11939

**Pull request merged**